src/linear_algebra/free_module_pid.lean

```diff
@@ -404,6 +608,109 @@ noncomputable def module.free_of_finite_type_torsion_free' [module.finite R M]
   Σ (n : ℕ), basis (fin n) R M :=
 module.free_of_finite_type_torsion_free module.finite.exists_fin.some_spec.some_spec
 
+/-- If `M` is finite free over a PID `R`, then any submodule `N` is free
+and we can find a basis for `M` and `N` such that the inclusion map is a diagonal matrix.
+
+This is a strengthening of `submodule.basis_of_pid_of_le`.
+-/
+theorem submodule.exists_smith_normal_form_of_le [fintype ι]
+  (b : basis ι R M) (N O : submodule R M) (N_le_O : N ≤ O) :
+  ∃ (n o : ℕ) (hno : n ≤ o) (bO : basis (fin o) R O) (bN : basis (fin n) R N) (a : fin n → R),
```

Comment:
Does it make sense to define a `SNF` predicate, and then use that as the conclusion of this theorem?

Reply:
Thinking about how I'm going to apply it, I'll even turn it into a structure.

Reply:
I defined a `basis.smith_normal_form` structure for a pair of bases, for the whole module and a submodule, that are in SNF.